### PR TITLE
use build_nim.sh

### PR DESCRIPTION
**Pull request by stefantalpalaru**

### CODE REVIEW

**Comment by stefantalpalaru** [comment]
Ready for review.

Note that we're using the Nim building script from Nimbus' "devel" branch, with the assumption that it won't disappear any time soon: https://raw.githubusercontent.com/status-im/nimbus/devel/build_nim.sh

**Comment by zah** [comment]
On top of what @mratsim said, I would also mention that when new shell scripts are needed, it's worth considering if the same job can be done with a `nims` script. Shell scripts often require more arcane constructs and while we've managed to put together a cross-platform environment relying on mingw, the portability story will be better if gradually shift to Nim-only solutions.

**Comment by stefantalpalaru** [comment]
> when new shell scripts are needed, it's worth considering if the same job can be done with a nims script

I hear you, but nims is no match for the conciseness of shell scripts, nor for the decades of experience that went into refining their "builtin" APIs. You see the difference when you try to pass arguments to a \*.nims script and you find out that you're limited to valid compiler parameters: https://github.com/status-im/nimbus/blob/0432283ad7d7a84fe91926b4af5514a9a5f204c2/nimbus.nimble#L25

Nim scripting is more of a quick hack for a very small subset of package building/testing/installing tasks. If we want a Nim shell replacement we need a new design and, even then, it's hard to beat Bash at this game for non-trivial scenarios.

Even Perl and Python failed at this, so it's too big of a task for this flimsy API: https://nim-lang.org/docs/nimscript.html

In conclusion, my position is: pragmatism before advocacy.

**Comment by zah** [comment]
Hrm, I tried to use `confutils` in nimscript today and everything worked fine after fixing some minor issues:

#### test.nims
```nim
import
  confutils

cli do (someSetting: int):
  echo someSetting
```

```
$ nim test.nims --someSetting=10
Hint: used config file '/Users/zahary/nimbus/vendor/Nim/config/nim.cfg' [Conf]
Hint: used config file '/Users/zahary/nimbus/nim.cfg' [Conf]
Hint: used config file '/Users/zahary/nimbus/vendor/nim-confutils/test.nims' [Conf]
10
```

Did you meant that it's not possible to pass the normal options such as `verbosity` and `hints` when you invoke a nims file?

**Comment by stefantalpalaru** [comment]
> Did you meant that it's not possible to pass the normal options such as verbosity and hints when you invoke a nims file?

The opposite: you' can't pass "--foo" because it's not a valid compiler option. You can only pass "-d:foo".

**Comment by stefantalpalaru** [comment]
OK, I haven't tried placing those parameters after the \*.nims file. You have to admit it's weird, though. Specially when you also specify a task on the command line, which has to come before the script name.

**Comment by zah** [comment]
The approach above mimics what you would do with an interpreter such as `python` or `ruby`. These printed out hints seem to be the only weird part.

Anyway, it's up to you, but I reflected a bit on what you said. Usually, when I choose `bash` over `ruby`, it's not because `bash` is more "refined", but rather because I don't want to introduce a whole new dependency on `ruby` (with nimscript, we don't have this drawback). 

Otherwise, the programming experience from a proper language is always superior for me than using `bash`. The text-based substitution paradigm starts to hurt as soon as you need to manipulate lists or more complicated structures and very often it starts to hurt so bad that you end up re-writing the whole thing from scratch in a real programming language (so, every time you start a script in bash, you are exposing yourself to such a risk).

**Comment by stefantalpalaru** [comment]
> the programming experience from a proper language is always superior for me when compared to shell scripting

Even when you take maintenance into consideration?

If we can agree that, for regular maintenance, difficulty is directly proportional to the number of lines of code, imagine writing fewer lines in Python or Ruby for something like: https://github.com/status-im/nimbus/blob/4c29a519e5a48f0278abf987be4e5e0c708bfa94/build_rocksdb.sh#L13-L16

And this is the more readable shell scripting style, with `if/then` instead of `&&` chains, so it's not code golf.

**Comment by zah** [comment]
YMMV, but all programming languages allow you to create abstractions that will directly map to the corresponding statements written in bash:

Ruby example:
```ruby
ROCKSDBVER = ENV["ROCKSDBVER"]
CI_CACHE = ENV["CI_CACHE"]

if not CI_CACHE.empty? or 
   Dir["#{CI_CACHE}/lib/librocksdb"].any? { |path| path.includes? ROCKSDBVER }
  %x{sudo cp -a "$CI_CACHE"/* /usr/local/}
  exit 0
end
```

In Nim, the same is possible, but I'd agree that a lot of the APIs are still not easy to use in NimScript.

**Comment by stefantalpalaru** [comment]
> `Dir["#{CI_CACHE}/lib/librocksdb"].any? { |path| path.includes? ROCKSDBVER }`

This is missing the globbing. It should be:

`Dir["#{CI_CACHE}/lib/librocksdb*"].any? { |path| path.include? ROCKSDBVER }`

> `%x{sudo cp -a "$CI_CACHE"/* /usr/local/}`

This looks like it would launch a shell process just for that one line. A native solution would look more like this: https://docs.ruby-lang.org/en/2.6.0/Shell.html
